job-runner/chain/reaper.go

```diff
@@ -0,0 +1,589 @@
+	RMCTries     int
+	RMCRetryWait time.Duration
+	DoneChan     chan struct{}
+	DoneJobChan  chan proto.Job
+	RunJobChan   chan proto.Job
+	StopTimeout  time.Duration
+}
+
+// Make a JobReaper for use on a running job chain.
+func (f *ChainReaperFactory) MakeRunning(runningJobs map[string]struct{}) JobReaper {
+	return &RunningChainReaper{
+		reaper: reaper{
+			chain:             f.Chain,
+			chainRepo:         f.ChainRepo,
+			rmc:               f.RMClient,
+			logger:            f.Logger,
+			finalizeTries:     f.RMCTries,
+			finalizeRetryWait: f.RMCRetryWait,
+			doneChan:          f.DoneChan,
+			doneStopping:      make(chan struct{}),
+			doneJobChan:       f.DoneJobChan,
+			stopChan:          make(chan struct{}),
+			running:           runningJobs,
+		},
+		runJobChan: f.RunJobChan,
+	}
+}
+
+// Make a JobReaper for use on a job chain being suspended.
+func (f *ChainReaperFactory) MakeSuspended(runningJobs map[string]struct{}) JobReaper {
+	return &SuspendedChainReaper{
+		reaper: reaper{
+			chain:             f.Chain,
+			chainRepo:         f.ChainRepo,
+			rmc:               f.RMClient,
+			logger:            f.Logger,
+			finalizeTries:     f.RMCTries,
+			finalizeRetryWait: f.RMCRetryWait,
+			doneChan:          f.DoneChan,
+			doneStopping:      make(chan struct{}),
+			doneJobChan:       f.DoneJobChan,
+			stopChan:          make(chan struct{}),
+			running:           runningJobs,
+		},
+		stopTimeout: f.StopTimeout,
+	}
+}
+
+// Make a JobReaper for use on a job chain being stopped.
+func (f *ChainReaperFactory) MakeStopped(runningJobs map[string]struct{}) JobReaper {
+	return &StoppedChainReaper{
+		reaper: reaper{
+			chain:             f.Chain,
+			chainRepo:         f.ChainRepo,
+			rmc:               f.RMClient,
+			logger:            f.Logger,
+			finalizeTries:     f.RMCTries,
+			finalizeRetryWait: f.RMCRetryWait,
+			doneChan:          f.DoneChan,
+			doneStopping:      make(chan struct{}),
+			doneJobChan:       f.DoneJobChan,
+			stopChan:          make(chan struct{}),
+			running:           runningJobs,
+		},
+		stopTimeout: f.StopTimeout,
+	}
+}
+
+// -------------------------------------------------------------------------- //
+
+// Job Reaper for running chains.
+type RunningChainReaper struct {
+	reaper
+	runJobChan chan proto.Job // enqueue next jobs to run here
+}
+
+// Run reaps jobs when they finish running. For each job reaped, if...
+// - chain is done: save final state + send to RM.
+// - job failed:    retry sequence if possible.
+// - job completed: prepared subsequent jobs and enqueue if runnable.
+func (r *RunningChainReaper) Run() {
+REAPER:
+	for {
+		select {
+		case job := <-r.doneJobChan:
+			delete(r.running, job.Id)
+			r.Reap(job)
+		case <-r.stopChan:
+			close(r.doneStopping)
+			return
+		default:
+			// If there were no jobs done running, and there are no jobs currently
+			// running, then there must be no more jobs to run - the chain is done.
+			if len(r.running) == 0 {
```

Comment:
This creates an infinite, uncontrolled loop at times. For example, if we start a long-running job, this `for` loop as fast as possible, hitting the default case. That's too inefficient. Need a more controlled way to know when done.